### PR TITLE
ARROW-4787: [C++] Add support for Null in MemoTable and related kernels

### PR DESCRIPTION
**Pull request by fsaintjacques**

### CODE REVIEW

**Comment by pitrou** [comment]
@fsaintjacques Can you fix the Python tests?

File: cpp/src/arrow/util/hashing.h [hunk 1]
```diff
@@ -698,6 +808,23 @@ struct HashTraits<T, enable_if_fixed_size_binary<T>> {
...
+                                           int64_t start_offset, int64_t* null_count,
+                                           std::shared_ptr<Buffer>* validity_bitmap) {
+  int64_t dict_length = static_cast<int64_t>(memo_table.size()) - start_offset;
+  int64_t null_index = memo_table.GetNull();
+  if (null_index != kKeyNotFound && null_index >= start_offset) {
+    null_index -= start_offset;
+    *null_count = 1;
+    RETURN_NOT_OK(
+        internal::BitmapAllButOne(pool, dict_length, null_index, validity_bitmap));
+  }
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/hashing.h]: Same for `*null_count = 0`.

File: cpp/src/arrow/util/hashing.h [hunk 2]
```diff
@@ -698,6 +808,23 @@ struct HashTraits<T, enable_if_fixed_size_binary<T>> {
   using MemoTableType = BinaryMemoTable;
 };
 
+template <typename MemoTableType>
+static inline Status ComputeValidityBitmap(MemoryPool* pool,
```
Comment (pitrou) [inline cpp/src/arrow/util/hashing.h]: We usually call it "null bitmap" ;-)

**Review by pitrou — COMMENTED** [review]
Looks good on the principle. The dict-encode behaviour is an open question. Also, tests need to be fixed on the Python side.

File: cpp/src/arrow/util/hashing.h [hunk 3]
```diff
@@ -35,6 +34,7 @@
 #include "arrow/array.h"
 #include "arrow/buffer.h"
 #include "arrow/builder.h"
+#include "arrow/compute/kernels/util-internal.h"
```
[the remark this answers is not in the document]
Comment (fsaintjacques) [reply cpp/src/arrow/util/hashing.h]: This is a leftover from a refactor I did. I'll remove it.

File: cpp/src/arrow/compute/kernels/hash-test.cc [hunk 4]
```diff
@@ -160,8 +161,8 @@ TYPED_TEST(TestHashKernelPrimitive, DictEncode) {
   using T = typename TypeParam::c_type;
   auto type = TypeTraits<TypeParam>::type_singleton();
   CheckDictEncode<TypeParam, T>(&this->ctx_, type, {2, 1, 2, 1, 2, 3},
-                                {true, false, true, true, true, true}, {2, 1, 3}, {},
-                                {0, 0, 0, 1, 0, 2});
+                                {true, false, true, true, true, true}, {2, 0, 1, 3},
+                                {1, 0, 1, 1}, {0, 0, 0, 2, 0, 3});
```
[the remark this answers is not in the document]
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/hash-test.cc]: Once this is merged, I need to open a discussion on the mailing list as the format spec doesn't mandate any of this.

File: cpp/src/arrow/util/hashing.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (fsaintjacques) [reply cpp/src/arrow/util/hashing.h]: It is done by the caller, but I will also explicit it in the function (compiler should optimize it away).

File: cpp/src/arrow/compute/kernels/hash-test.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/hash-test.cc]: Right. But for the time being, we should probably choose one or the other. I don't think it makes sense to add a null to the dict values if it's not referred to by the indices.
(I think for compatibility we should perhaps keep the current behaviour @wesm )

[the remark this answers is not in the document]
Comment (fsaintjacques) [reply cpp/src/arrow/compute/kernels/hash-test.cc]: I'm working on keeping the current behavior (which fill fix the python test).

File: cpp/src/arrow/util/hashing.h (hunk 2 above)
Comment (wesm) [reply cpp/src/arrow/util/hashing.h]: "validity bitmap" is used in some other places (e.g. in the Arrow spec itself), it's my fault for using the "null bitmap" terminology in the first place in this codebase

File: cpp/src/arrow/compute/kernels/hash-test.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/compute/kernels/hash-test.cc]: I don't think this is a specification-related matter. The data is simply data

**Comment by fsaintjacques** [comment]
@pitrou I reverted the DictEncode behaviour, thus no need to change the python tests (and also pandas categorical conversion). While doing so, I refactored the HashKernel a bit since there's 2 bit provided at compile time. 

I'm still not satisfied, but it'll do for now. I believe we'll have to break the HashKernel/Action implementation into specific kernels soon.

**Comment by pitrou** [comment]
It looks like the Python tests still need a fix (see AppVeyor results).

**Review by pitrou — APPROVED** [review]
+1. Thank you @fsaintjacques !

**Pull request closed without merging**